### PR TITLE
Jmpi add matching rule to audit trail

### PR DESCRIPTION
**Pull request by NyashaMuusha**
### 1. Description of changes
Modified database audit event to store event type and event data to make it more generic and reusable
De-serialized data based on event type
Calculated linking rule from golden record and interaction record and added value to audit event
Modified API response for audit trail to reflect new changes
Added matching type column on audit trail table in UI

### 2. How to test/configure
Navigate to http://localhost:3000/browse-records
Double click on a record
AuditTrailMatchingRule

### CODE REVIEW

**Comment by walisc** [comment]
Hi @NyashaMuusha . Could you update your branch with `dev`. i.e pull `dev` into your branch, and push again. p.s All you checks should be passing.

File: JeMPI_Apps/JeMPI_UI/src/utils/constants.ts [hunk 1]
```diff
@@ -92,5 +93,24 @@ export const AUDIT_TRAIL_COLUMNS: GridColDef[] = [
...
+    headerName: 'Matching Type',
+    valueFormatter: ({ value }) => extractMatchingRule(value),
+    sortable: false,
+    disableColumnMenu: true,
+    headerClassName: 'super-app-theme--header',
+    flex: 1
   }
 ]
+
+const extractMatchingRule = (score: number) => {
```
Comment (walisc) [inline JeMPI_Apps/JeMPI_UI/src/utils/constants.ts:108]: Hi @NyashaMuusha ,

As mention in the previous PR, score does not determine if something is matched Deterministically or Probablistically. It just happens that when we link Deterministically the score it 1.0 (i.e although quite rare, you could have a scenario where you have a score of 1.0 but linked Probablistically). 

What you want to try do is something like in the `calcNormalizedScore` method in `JeMPI_Apps/JeMPI_Linker/src/main/java/org/jembi/jempi/linker/backend/LinkerUtils.java`

Once you you obtain if it is linking Deterministically or Probablistically, update the `WorkCandidate` in `JeMPI_Apps/JeMPI_Linker/src/main/java/org/jembi/jempi/linker/backend/LinkerDWH.java` with these detail (in addition to the score)

When it calls `createInteractionAndLinkToExistingGoldenRecord` you will need to pass through these details

**Review by walisc — CHANGES_REQUESTED** [review]

**Review by walisc — CHANGES_REQUESTED** [review]

File: JeMPI_Apps/JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/LibMPI.java [hunk 2]
```diff
@@ -72,6 +76,27 @@ private void sendAuditEvent(
 
    }
 
+   private void sendAuditEvent(
+           final String interactionID,
+           final String goldenID,
+           final String event,
+           final Float score) {
+      topicAuditEvents.produceAsync(goldenID,
```
[the remark this answers is not in the document]
Comment (NyashaMuusha) [reply JeMPI_Apps/JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/LibMPI.java]: f0f694b

File: JeMPI_Apps/JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/LibMPI.java [hunk 3]
```diff
@@ -54,22 +57,34 @@ public LibMPI(
       client = new LibPostgresql(URL, USR, PSW);
    }
 
-   private void sendAuditEvent(
```
Comment (walisc) [inline JeMPI_Apps/JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/LibMPI.java:57]: Can you make the shared as discussed

**Review by walisc — CHANGES_REQUESTED** [review]

File: JeMPI_Apps/JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/LibMPI.java (hunk 3 above)
Comment (NyashaMuusha) [reply JeMPI_Apps/JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/LibMPI.java:57]: [c8d1dbf](https://github.com/jembi/JeMPI/pull/187/commits/c8d1dbf71d082a82b7c09e89975720483ef626a5)

File: JeMPI_Apps/JeMPI_Bootstrapper/src/main/resources/data/postgres/audit-schema.sql [hunk 4]
```diff
@@ -5,7 +5,9 @@ CREATE TABLE IF NOT EXISTS audit_trail (
     interactionID  VARCHAR(64),
     goldenID       VARCHAR(64),
     event          VARCHAR(256),
+    eventData      VARCHAR(256),
```
[the remark this answers is not in the document]
Comment (NyashaMuusha) [reply JeMPI_Apps/JeMPI_Bootstrapper/src/main/resources/data/postgres/audit-schema.sql]: [c8d1dbf](https://github.com/jembi/JeMPI/pull/187/commits/c8d1dbf71d082a82b7c09e89975720483ef626a5)

File: JeMPI_Apps/JeMPI_Bootstrapper/src/main/resources/data/postgres/audit-schema.sql [hunk 5]
```diff
@@ -2,10 +2,9 @@ CREATE TABLE IF NOT EXISTS audit_trail (
     id             UUID         NOT NULL DEFAULT gen_random_uuid(),
     insertedAt     TIMESTAMP    NOT NULL DEFAULT now(),
     createdAt      TIMESTAMP    NOT NULL,
-    interactionID  VARCHAR(64),
-    goldenID       VARCHAR(64),
-    event          VARCHAR(256),
+    eventType      VARCHAR(256),
+    eventData      TEXT,
```
Comment (walisc) [inline JeMPI_Apps/JeMPI_Bootstrapper/src/main/resources/data/postgres/audit-schema.sql:6]: Could we JSONB here, and filter as appropriate below. In `JeMPI_Apps/JeMPI_LibAPI/src/main/java/org/jembi/jempi/libapi/PsqlAuditTrail.java`

File: JeMPI_Apps/JeMPI_LibShared/src/main/java/org/jembi/jempi/shared/utils/AuditTrailUtil.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import org.jembi.jempi.shared.kafka.MyKafkaProducer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jembi.jempi.shared.models.AuditEvent;
+import org.jembi.jempi.shared.models.GlobalConstants;
+
+import static org.jembi.jempi.shared.utils.AppUtils.OBJECT_MAPPER;
+
+
+public final class AuditTrailUtil {
```
Comment (walisc) [inline JeMPI_Apps/JeMPI_LibShared/src/main/java/org/jembi/jempi/shared/utils/AuditTrailUtil.java:16]: Could you rename this to something like `AuditTrailBridge` as Util classes don't usually get instantiated

**Review by walisc — CHANGES_REQUESTED** [review]

**Review by walisc — APPROVED** [review]
Nice @NyashaMuusha ! Looking good!  :raised_hands:

**Comment by walisc** [comment]
p.s you just need to pull `dev`, resolve the conflicts, and push 👍🏽

**Pull request merged**